Auxiliary file download

pdurbin opened the pull request:
**What this PR does / why we need it**:

Adds the ability to download auxiliary files from the web interface. Also, other cleanup was done.

**Which issue(s) this PR closes**:

Closes #7400

**Special notes for your reviewer**:

I talked to @djbrooke about trimming down the changes he made but ultimately didn't because I wasn't sure what to take out or how to best word things. I think it's fine to leave it all in but I'm also fine with some of it being removed.

**Suggestions on how to test this**:

Use AuxiliaryFilesIT or your own methods to add some auxiliary files. Here are the changes to watch for (from the release note):

Auxiliary Files can now be downloaded from the web interface.

- Aux files uploaded as type=DP appear under "Differentially Private Statistics" under file level download. The rest appear under "Other Auxiliary Files".

In addition, related changes were made, including the following:

- New tooltip over the lock indicating if you have been granted access to a restricted file or not.
- When downloading individual files, you will see "Restricted with Access Granted" or just "Restricted" (followed by "Users may not request access to files.") as appropriate.
- When downloading individual files, instead of "Download" you should expect to see the file type such as "JPEG Image" or "Original File Format" if the type is unknown.
- Downloaded aux files now have a file extension if it can be determined.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

Yes. Mockups at https://docs.google.com/presentation/d/1XuX4qStdHjkFf4g1d1vKxxb85kuuUSEzdKN41dNM2mM/

Here's a screenshot showing the DP files:

<img width="1285" alt="Screen Shot 2021-03-26 at 2 16 23 PM" src="https://user-images.githubusercontent.com/21006/112675715-e02b1080-8e3d-11eb-92d2-c997fb57668b.png">

And another showing "other" files:

<img width="691" alt="Screen Shot 2021-03-26 at 2 27 22 PM" src="https://user-images.githubusercontent.com/21006/112676980-888da480-8e3f-11eb-985c-588378f04fa6.png">


**Is there a release notes update needed for this change?**:

Added.

**Additional documentation**:

Included.

djbrooke commented:
Thanks @pdurbin for the PR and for the note about the docs. I took another read through and I think they are fine for now, since we flag this as experimental.

pdurbin commented:
> * If the tag for an aux file starts with "dp" it will be displayed under "Differentially Private Statistics" under file level download.

Over at https://github.com/opendp/opendp-ux/issues/135#issuecomment-808459070 @TaniaSchlatter asked if this could be changed to "dp_" (adding an underscore) and this got a thumbs up from @raprasad. 

Yes, this is certainly possible. Let's see what other changes are requested and do them at once.

pdurbin commented:
Instead of "starts with dp" we decided to add a proper "type" parameter: d368a85

The downloaded files never had any file extensions. This was fixed in 5c67f95.

Ready for more review.

pdurbin commented:
In a1e852a I fixed a bug where I noticed that a user with no special access could download an aux file while it was still in draft. I also added some more tests having to do with permissions and the `isPublic` boolean. I also clarified the docs.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/AuxiliaryFile.java:
```diff
@@ -44,6 +48,29 @@
+    @Enumerated(EnumType.STRING)
+    private Type type;
+
+    public enum Type {
```
rather than an enum, can we make this a String? the way to handle the friendly text (for now) would be to just return the value for OTHER for anything other than DP. (eventually this would have to be more robust and first check to see if a bundle value exists before defaulting to other.

pdurbin replied:
Done in 9e5f39089.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/AuxiliaryFile.java:
```diff
@@ -115,6 +142,25 @@ public String getChecksum() {
+     */
+    public String getTypeFriendly() {
+        if (this.type == null) {
+            return null;
```
should null type, return other? (see the other note for a more generally suggested change, anyway)

pdurbin replied:
This was rewritten as part of 9e5f39089. Unspecified (null) type gets grouped as "other" in the UI.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/AuxiliaryFileServiceBean.java:
```diff
@@ -114,4 +119,32 @@ public AuxiliaryFile lookupAuxiliaryFile(DataFile dataFile, String formatTag, St
         }
     }
 
+    public List<AuxiliaryFile> findAuxiliaryFiles(DataFile dataFile) {
+        TypedQuery query = em.createQuery("select object(o) from AuxiliaryFile as o where o.dataFile.id = :dataFileId", AuxiliaryFile.class);
+        query.setParameter("dataFileId", dataFile.getId());
+        return query.getResultList();
+    }
+
+    public List<String> findAuxiliaryFileTypes(DataFile dataFile) {
+        Query query = em.createNativeQuery("select distinct type from auxiliaryfile where datafile_id = ?1");
+        query.setParameter(1, dataFile.getId());
+        return query.getResultList();
+    }
+
+    public List<AuxiliaryFile> findAuxiliaryFilesByType(DataFile dataFile, String typeString) {
+        TypedQuery query = em.createQuery("select object(o) from AuxiliaryFile as o where o.dataFile.id = :dataFileId and o.type like :type", AuxiliaryFile.class);
+        query.setParameter("dataFileId", dataFile.getId());
+        query.setParameter("type", typeString);
+        return query.getResultList();
+    }
+
```
can we make these NamedQueries under AuxiliaryFile?

pdurbin replied:
Done in 3441758c9. I also moved existing queries.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/api/Access.java:
```diff
@@ -584,7 +587,13 @@ public DownloadInstance tabularDatafileMetadataAux(@PathParam("fileId") String f
             if (auxFile.getIsPublic()) {
-                publiclyAvailable = true;
+                // Require at least one published version for isPublic to be true.
+                for (FileMetadata fm : df.getFileMetadatas()) {
```
is it possible to check the if the datafile is published instead? Might be better than going through all the versions? (I can remember if we have a file.isPublished() method, or maybe a file.getPublishicationDate()? and check for !null?

pdurbin replied:
Fixed in f55b36862.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/api/DownloadInstanceWriter.java:
```diff
@@ -238,7 +241,18 @@ public void writeTo(DownloadInstance di, Class<?> clazz, Type type, Annotation[]
                     InputStreamIO auxStreamIO = new InputStreamIO(storageIO.getAuxFileAsInputStream(auxTag), auxFileSize);
-                    auxStreamIO.setFileName(storageIO.getFileName() + "." + auxTag);
+                    String fileExtension = "";
+                    String contentType = di.getAuxiliaryFile().getContentType();
```
minor tweak, but this looks like it could be good as a separate method to make this code more streamlined.

pdurbin replied:
Fixed in 9b852de13.

scolapasta commented on src/main/java/propertyFiles/Bundle.properties:
```diff
@@ -1585,6 +1585,8 @@ file.metaData.checksum.copy=Click to copy
 file.metaData.dataFile.dataTab.unf=UNF
 file.metaData.dataFile.dataTab.variables=Variables
 file.metaData.dataFile.dataTab.observations=Observations
+file.metaData.access.restricted=File Access: Restricted
+file.metaData.access.granted=File Access: Restricted with Access Granted
```
Minor, but it feels these could be broken up - one bundle for "File Access" and another (two) for "Restricted" and "Restricted with Access Granted". The reason is because further down in the bundle "Restricted" and "Restricted with Access Granted" and defined again for the download button. So if a change is made, it would have to be on both places (assuming that the same text is wanted in both). And I'm sure there are other places we use "Restricted" and "Restricted with Access Granted". All that said, that could maybe be a cleanup for some other time, but wanted to mention here in case we wanted to get a head start.

scolapasta replied:
For example the bundle also has:
restricted=Restricted
restrictedaccess=Restricted Access

(not sure if the 2nd one is the equivalent of "Restricted with Access Granted" but feels like it would be)

pdurbin replied:
Fixed in 6f8ddc933. I added "Restricted with Access Granted" as its own entry because I don't know its relationship to "Restricted Access".

scolapasta replied:
I added a new note for this (after a quick check of where "Restricted Access" was used. (I should've probably just replied here - sorry.

scolapasta commented on src/main/webapp/dataset.xhtml:
```diff
@@ -37,8 +37,9 @@
+            <!-- NOTE removed fileAccessRequest logic in order to always display the file access btn -->
             <ui:param name="showAccessFileButtonGroup" value="#{fileDownloadHelper.canDownloadFile(fileMetadata)
-                                                                or (!fileDownloadHelper.canDownloadFile(fileMetadata) and DatasetPage.dataset.fileAccessRequest)
+                                                                or (!fileDownloadHelper.canDownloadFile(fileMetadata))
```
this rendered should be cleaned up.

pdurbin replied:
Fixed in 7903a8f5d.

scolapasta commented on src/main/webapp/file.xhtml:
```diff
@@ -18,8 +18,9 @@
             <ui:param name="showMessagePanel" value="#{true}"/>
+            <!-- NOTE removed fileAccessRequest logic in order to always display the file access btn -->
             <ui:param name="showAccessFileButtonGroup" value="#{fileDownloadHelper.canDownloadFile(FilePage.fileMetadata)
-                                                                  or (!fileDownloadHelper.canDownloadFile(FilePage.fileMetadata) and FilePage.fileMetadata.dataFile.owner.fileAccessRequest)
```
another spot for some rendering clean up

pdurbin replied:
Fixed in 7903a8f5d.

scolapasta commented on src/test/java/edu/harvard/iq/dataverse/api/AccessIT.java:
```diff
@@ -103,6 +103,11 @@ public static void setUp() throws InterruptedException {
         createDatasetResponse.prettyPrint();
         datasetId = JsonPath.from(createDatasetResponse.body().asString()).getInt("data.id");
+
+        // TODO: Get this working.
```
is this todo important before this gets merged?

pdurbin replied:
I just removed the TODO in 81628864c. Not sure why I put it there because "allow access requests" seems to be working fine.

scolapasta reviewed (changes requested):
Good stuff! Made some smaller suggestions for cleanup (there's a bundle one, which could be bumped BUT it relates to ones we're adding now, so might be good to tackle now. Biggest change requested is witching away from an enum for types of Aux Files.

pdurbin commented on doc/sphinx-guides/source/developers/aux-file-support.rst:
```diff
@@ -34,3 +35,5 @@ formatTag and formatVersion (if applicable) associated with the auxiliary file:
 
   curl "$SERVER_URL/api/access/datafile/$FILE_ID/$FORMAT_TAG/$FORMAT_VERSION"
+
+It is also possible to download the auxiliary file from the web interface. Auxiliary files that have a tag beginning with "dp" appear under "Differentially Private Statistics". The rest appear under "Other Auxiliary Files".
```
Done in 9e5f39089.

pdurbin commented on doc/sphinx-guides/source/developers/aux-file-support.rst:
```diff
@@ -1,11 +1,11 @@
 Adding an Auxiliary File to a Datafile
 --------------------------------------
-To add an auxiliary file, specify the primary key of the datafile (FILE_ID), and the formatTag and formatVersion (if applicable) associated with the auxiliary file. There are two form parameters. "Origin" specifies the application/entity that created the auxiliary file, an "isPublic" controls access to downloading the file. If "isPublic" is true, any user can download the file, else, access autho ...
+To add an auxiliary file, specify the primary key of the datafile (FILE_ID), and the formatTag and formatVersion (if applicable) associated with the auxiliary file. There are multiple form parameters. "Origin" specifies the application/entity that created the auxiliary file, and "isPublic" controls access to downloading the file. If "isPublic" is true, any user can download the file if the datase ...
```
Done in 9e5f39089.

scolapasta commented on doc/release-notes/7400-opendp-download.md:
```diff
@@ -0,0 +1,10 @@
+Auxiliary Files can now be downloaded from the web interface.
+
+- Aux files uploaded as type=DP appear under "Differentially Private Statistics" under file level download. The rest (type=OTHER) appear under "Other Auxiliary Files".
```
it's now more than just type=OTHER

pdurbin replied:
Right, and type=OTHER isn't a thing anymore, so I removed this reference to it in 20b438201.

scolapasta commented on doc/sphinx-guides/source/developers/aux-file-support.rst:
```diff
@@ -1,11 +1,11 @@
 Adding an Auxiliary File to a Datafile
 --------------------------------------
-To add an auxiliary file, specify the primary key of the datafile (FILE_ID), and the formatTag and formatVersion (if applicable) associated with the auxiliary file. There are two form parameters. "Origin" specifies the application/entity that created the auxiliary file, an "isPublic" controls access to downloading the file. If "isPublic" is true, any user can download the file, else, access autho ...
+To add an auxiliary file, specify the primary key of the datafile (FILE_ID), and the formatTag and formatVersion (if applicable) associated with the auxiliary file. There are multiple form parameters. "Origin" specifies the application/entity that created the auxiliary file, and "isPublic" controls access to downloading the file. If "isPublic" is true, any user can download the file if the datase ...
```
Suggested wording:
The "type" parameter is used to group similar auxiliary files in the UI. Currently, for type "DP", they are appear under "Differentially Private Statistics", while all other auxiliary files appear under "Other Auxiliary Files".

pdurbin replied:
Fixed in 5dc723420.

scolapasta replied:
After seeing it again, I made a minor tweak (and committed it).

scolapasta commented on doc/sphinx-guides/source/developers/aux-file-support.rstnew:
```diff
@@ -0,0 +1,39 @@
+
+Adding an Auxiliary File to a Datafile
+--------------------------------------
+To add an auxiliary file, specify the primary key of the datafile (FILE_ID), and the formatTag and formatVersion (if applicable) associated with the auxiliary file. There are multiple form parameters. "Origin" specifies the application/entity that created the auxiliary file, and "isPublic" controls access to downloading the file. If "isPublic" is true, any user can download the file if the datase ...
```
can we not use "Other" as a type and just have it be blank if they don't want to specify?

pdurbin replied:
My bad, this was an old version of the aux file page that I accidentally committed. Removed in 51cffe8b1.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/AuxiliaryFile.java:
```diff
@@ -1,20 +1,40 @@
+            query = "select object(o) from AuxiliaryFile as o where o.dataFile.id = :dataFileId"),
+    @NamedQuery(name = "AuxiliaryFile.findAuxiliaryFilesByType",
+            query = "select object(o) from AuxiliaryFile as o where o.dataFile.id = :dataFileId and o.type like :type"),
+    @NamedQuery(name = "AuxiliaryFile.findOtherAuxiliaryFiles",
```
This is the same as the previous query, no?

pdurbin replied:
Good catch. Fixed in b132580b8.

scolapasta commented on src/main/webapp/dataset.xhtml:
```diff
@@ -37,10 +37,10 @@
-                                                                or (!fileDownloadHelper.canDownloadFile(fileMetadata) and DatasetPage.dataset.fileAccessRequest)
-                                                                and (!DatasetPage.workingVersion.deaccessioned
-                                                                or (DatasetPage.workingVersion.deaccessioned and permissionsWrapper.canIssueUpdateDatasetCommand(FilePage.fileMetadata.datasetVersion.dataset)))}"/>
+            <!-- NOTE removed fileAccessRequest logic in order to always display the file access btn -->
```
Often we put comments like this within a ui:remove so it doesn't show in browser under "view source"

pdurbin replied:
That's true. I didn't add this one. Future cleanup?

scolapasta replied:
If you feel like adding now, go for it; if not it's not critical (the comment doesn't "leak" anything it shouldn't).

pdurbin replied:
Leaving this as future cleanup is fine for me.

scolapasta commented on src/main/java/propertyFiles/Bundle.properties:
```diff
@@ -11,6 +11,7 @@ file=File
 public=Public
 restricted=Restricted
 restrictedaccess=Restricted Access
+restrictedWithAccessGranted=Restricted with Access Granted
```
I checked is only uses as a title on an image lock of a file search card; so let's have consistency and just have the one (restrictedaccess=Restricted with Access Granted)

pdurbin replied:
Done in 47ef8f670. Now the search card is consistent with other places:

<img width="417" alt="Screen Shot 2021-04-16 at 11 17 07 AM" src="https://user-images.githubusercontent.com/21006/115047013-1b9d8580-9ea6-11eb-8c57-768ed96ce77a.png">

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/AuxiliaryFile.java:
```diff
@@ -1,20 +1,39 @@
+    @NamedQuery(name = "AuxiliaryFile.findAuxiliaryFiles",
+            query = "select object(o) from AuxiliaryFile as o where o.dataFile.id = :dataFileId"),
+    @NamedQuery(name = "AuxiliaryFile.findAuxiliaryFilesByType",
+            query = "select object(o) from AuxiliaryFile as o where o.dataFile.id = :dataFileId and o.type like :type"),
```
Just noticed this - why "like" and not straight "="

pdurbin replied:
No reason. Fixed in e1d5541fd.

scolapasta commented on src/main/java/propertyFiles/Bundle.properties:
```diff
@@ -1620,10 +1621,13 @@ file.ingestFailed.header=Upload Completed with Errors
+file.downloadBtn.format.rdata=RData
 file.downloadBtn.format.var=Variable Metadata
 file.downloadBtn.format.citation=Data File Citation
+file.downloadBtn.restricted=Restricted
+file.downloadBtn.restricted.with.access.granted=Restricted with Access Granted
```
We can now get rid of these and just use:
restricted=Restricted
restrictedaccess=Restricted with Access Granted

scolapasta reviewed (approved):
With the exception of deciding on the API (which we documented as a todo and also explicitly mentioned in the "experimental" section, I think this is good to go. Next up, QA!
